_overviews/overview/a-taste-of-scala.md

```diff
@@ -0,0 +1,1329 @@
+
+```scala
+def double(i: Int): Int = i * 2
+val a = List(1,2,3).map(i => double(i))
+val b = List(1,2,3).map(double)
+```
+
+>If you haven’t seen the `map` method before, it applies a given function to every element in a list, yielding a new list that contains the new values.
+
+
+### Higher-order functions
+
+A “higher order function” (HOF) is a function that can take a function as an input parameter, or returns a function as a return value. HOFs are extremely common in Scala, such as being used throughout the standard libraries. These examples show how to pass methods and anonymous functions into the functional methods of Scala collections:
+
+```scala
+// a sample list
+val nums = (1 to 10).toList   // List(1, 2, 3, 4, 5, 6, 7, 8, 9, 10)
+
+// sample methods
+def isEven(i: Int): Boolean = i % 2 == 0
+def double(i: Int): Int = i * 2
+
+// pass in a method or function
+val a = nums.filter(isEven)   // List(2, 4, 6, 8, 10)
+val b = nums.map(double)      // List(2, 4, 6, 8, 10, 12, 14, 16, 18, 20)
+
+// pass in a lambda
+val c = nums.filter(_ % 2 == 0)   // List(2, 4, 6, 8, 10)
+val d = nums.map(_ * 2)           // List(2, 4, 6, 8, 10, 12, 14, 16, 18, 20)
+
+// methods can be chained as needed
+val e = nums.filter(_ > 3)        // List(40, 50, 60)
+            .filter(_ < 7)
+            .map(_ * 10)
+```
+
+In addition to HOFs used all throughout the standard library, you can easily create your own HOFs. This is shown in the Reference documentation.
+
+<!--
+### Functions are objects
+- add this section?
+- TODO: show the types of functions?
+    - see my 'val function vs def method page'
+-->
+
+
+## Extension methods
+
+<!-- TODO: i may not be showing the latest syntax -->
+*Extension methods* let you add new methods to closed classes. For instance, if you want to add a `hello` method to the `String` class, just create an extension method:
+
+<!-- TODO: scalafiddle -->
+```scala
+def (s: String) hello: String = s"Hello, ${s.capitalize}"
+
+"world".hello    // "Hello, world"
+"friend".hello   // "Hello, friend"
+```
+
+The comments in this code explain the extension method syntax:
+
+````
+def (s: String) hello: String = s"Hello, ${s.capitalize}"
+    ----------  -----  ------   -------------------------
+        ^         ^      ^                ^
+    add method  method return         method body
+    to String    name   type
+      class
+````
+
+This next example shows how to add a `makeInt` method to the `String` class. Here, `makeInt` takes a parameter named `radix`. The code doesn’t account for possible string-to-integer conversion errors, but skipping that detail, the examples show how it works:
+
+<!-- TODO: scalafiddle -->
+```scala
+def (s: String) makeInt(radix: Int): Int = Integer.parseInt(s, radix)
+"1".makeInt(2)      // Int = 1
+"10".makeInt(2)     // Int = 2
+"100".makeInt(2)    // Int = 4
+```
+
+
+
+## Collections classes
```

Comment:
I’m not sure we should go into these details here. I would simply say that Scala provides a rich set of collections, which have a rich set of methods, and show a few examples, but without showing the tables listing the collection classes and the collection methods.

Reply:
I agree, those are too many details here. I’ll move those to the later Collections section. That will definitely help to shorten this — thanks!